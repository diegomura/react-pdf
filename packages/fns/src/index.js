export { default as adjust } from './adjust';
export { default as asyncCompose } from './asyncCompose';
export { default as capitalize } from './capitalize';
export { default as castArray } from './castArray';
export { default as compose } from './compose';
export { default as dropLast } from './dropLast';
export { default as evolve } from './evolve';
export { default as get } from './get';
export { default as isNil } from './isNil';
export { default as last } from './last';
export { default as mapValues } from './mapValues';
export { default as matchPercent } from './matchPercent';
export { default as omit } from './omit';
export { default as pick } from './pick';
export { default as repeat } from './repeat';
export { default as reverse } from './reverse';
export { default as upperFirst } from './upperFirst';
