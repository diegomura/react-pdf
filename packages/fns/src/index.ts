export { default as adjust } from './adjust.js';
export { default as asyncCompose } from './asyncCompose.js';
export { default as capitalize } from './capitalize.js';
export { default as castArray } from './castArray.js';
export { default as compose } from './compose.js';
export { default as dropLast } from './dropLast.js';
export { default as evolve } from './evolve.js';
export { default as get } from './get.js';
export { default as isNil } from './isNil.js';
export { default as last } from './last.js';
export { default as mapValues } from './mapValues.js';
export { default as matchPercent } from './matchPercent.js';
export { default as omit } from './omit.js';
export { default as pick } from './pick.js';
export { default as repeat } from './repeat.js';
export { default as reverse } from './reverse.js';
export { default as upperFirst } from './upperFirst.js';
export { default as without } from './without.js';
export { default as parseFloat } from './parseFloat.js';
