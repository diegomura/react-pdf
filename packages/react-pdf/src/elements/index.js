import Text from './Text';
import View from './View';
import Page from './Page';
import Document from './Document';
import omit from 'lodash/fp/omit';

function createElement(type, props, root) {
  let instance;

  switch (type) {
    case 'ROOT':
      instance = new Document();
      break;
    case 'DOCUMENT':
      /*
       Since the Document instance is being created by the root element,
       all we do when the renderer gets to <Document /> is injecting
       it's props and return the root
      */
      root.setProps(omit('children', props));
      instance = root;
      break;
    case 'PAGE':
      instance = new Page(props, root);
      break;
    case 'TEXT':
      instance = new Text(props, root);
      break;
    case 'VIEW':
      instance = new View(props, root);
      break;
    default:
      instance = undefined;
  }

  return instance;
}

export { createElement };
