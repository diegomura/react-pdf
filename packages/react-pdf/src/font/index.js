import isUrl from 'is-url';
import standardFonts from './standard';
import { fetchFont } from '../utils/font';

let fonts = {};

const register = (src, { family, ...otherOptions }) => {
  fonts[family] = {
    src,
    loaded: false,
    data: null,
    ...otherOptions,
  };
};

const getRegisteredFonts = () => Object.keys(fonts);

const load = async (fontFamily, doc) => {
  const font = fonts[fontFamily];

  // We cache the font to avoid fetching it many time
  if (font && !font.data) {
    font.data = isUrl(font.src) ? await fetchFont(font.src) : font.src;
  }

  // If the font wasn't added to the document yet (aka. loaded), we do.
  // This prevents calling `registerFont` many times for the same font.
  // Fonts loaded state will be resetted after document is closed.
  if (font && !font.loaded) {
    font.loaded = true;
    doc.registerFont(fontFamily, font.data);
  }

  if (!font && !standardFonts.includes(fontFamily)) {
    throw new Error(
      `Font familiy not registered: ${fontFamily}. Please register it calling Font.register() method.`,
    );
  }
};

const reset = () => {
  for (const font in fonts) {
    if (fonts.hasOwnProperty(font)) {
      fonts[font].loaded = false;
    }
  }
};

const clear = () => {
  fonts = {};
};

export default {
  register,
  getRegisteredFonts,
  load,
  clear,
  reset,
};
