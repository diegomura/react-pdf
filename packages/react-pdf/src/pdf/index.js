/* global Blob */
import transformTree from './transformer';

const pdf = input => {
  const header = '%PDF-1.7\n';

  // Creates the footer of the PDF document
  function buildCrossReferenceTable(objects) {
    let offset = header.length - 1;

    const pointers = objects
      .map(object => {
        const pointer = `${('0000000000' + offset).slice(-10)} 00000 n`;
        offset += object.value.length;
        return pointer;
      })
      .join('\n');

    return 'xref\n' +
      `0 ${objects.length + 1}\n` +
      '0000000000 65535 f\n' +
      `${pointers}\n` +
      'trailer\n' +
      '<<\n' +
      `/Size ${objects.length}\n` +
      '/Root 1 0 R\n' +
      '>>' +
      'startxref\n' +
      `${offset}\n` +
      '%%EOF';
  }

  function buildDocument(input) {
    // Transforms the nodes trees into a more suitable data structure
    const objects = transformTree(input);

    // Builds the body and footer of the document
    const body = objects.map(object => object.value).join('\n\n');
    const footer = buildCrossReferenceTable(objects);

    return header + body + footer;
  }

  function toBlob() {
    return new Blob([buildDocument(input)], {
      type: 'application/pdf',
    });
  }

  function toBuffer() {
    return new Buffer(buildDocument(input));
  }

  return {
    toBuffer,
    toBlob,
  };
};

export default pdf;
