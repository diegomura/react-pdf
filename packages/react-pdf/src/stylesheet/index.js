import matchMedia from 'media-engine';
import transformStyles from './transformStyles';

const create = styles => styles;

const flatten = input => {
  if (!Array.isArray(input)) {
    input = [input];
  }

  const result = input.reduce((acc, style) => {
    if (style) {
      Object.keys(style).forEach(key => {
        if (style[key]) {
          acc[key] = style[key];
        }
      });
    }

    return acc;
  }, {});

  return result;
};

const resolveMediaQueries = (input, container) => {
  const result = Object.keys(input).reduce((acc, key) => {
    if (/@media/.test(key)) {
      return {
        ...acc,
        ...matchMedia({ [key]: input[key] }, container),
      };
    }

    return { ...acc, [key]: input[key] };
  }, {});

  return result;
};

const resolve = (styles, container) => {
  if (!styles) {
    return null;
  }

  styles = flatten(styles);
  styles = resolveMediaQueries(styles, container);

  return transformStyles(styles);
};

const absoluteFillObject = {
  position: 'absolute',
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

module.exports = {
  hairlineWidth: 1,
  create,
  resolve,
  flatten,
  absoluteFillObject,
};
