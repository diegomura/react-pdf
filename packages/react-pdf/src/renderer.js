'use strict';

import ReactFiberReconciler from 'react-reconciler';
import emptyObject from 'fbjs/lib/emptyObject';
import { createElement } from './elements';

const PDFRenderer = ReactFiberReconciler({
  appendInitialChild(parentInstance, child) {
    if (parentInstance.appendChild) {
      parentInstance.appendChild(child);
    } else {
      parentInstance.document = child;
    }
  },

  createInstance(type, props, internalInstanceHandle) {
    return createElement(type, props, internalInstanceHandle);
  },

  createTextInstance(text, rootContainerInstance, internalInstanceHandle) {
    return text;
  },

  finalizeInitialChildren(domElement, type, props) {
    return false;
  },

  getPublicInstance(instance) {
    return instance;
  },

  prepareForCommit() {
    // Noop
  },

  prepareUpdate(domElement, type, oldProps, newProps) {
    return true;
  },

  resetAfterCommit() {
    // Noop
  },

  resetTextContent(domElement) {
    // Noop
  },

  getRootHostContext() {
    return emptyObject;
  },

  getChildHostContext() {
    return emptyObject;
  },

  shouldSetTextContent(type, props) {
    return false;
  },

  now: () => {},

  useSyncScheduling: true,

  mutation: {
    appendChild(parentInstance, child) {
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        parentInstance.document = child;
      }
    },

    appendChildToContainer(parentInstance, child) {
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        parentInstance.document = child;
      }
    },

    insertBefore(parentInstance, child, beforeChild) {
      // noob
    },

    insertInContainerBefore(parentInstance, child, beforeChild) {
      // noob
    },

    removeChild(parentInstance, child) {
      parentInstance.removeChild(child);
    },

    removeChildFromContainer(parentInstance, child) {
      if (parentInstance.removeChild) {
        parentInstance.removeChild(child);
      }
    },

    commitTextUpdate(textInstance, oldText, newText) {
      textInstance = newText;
    },

    commitMount(instance, type, newProps) {
      // Noop
    },

    commitUpdate(instance, updatePayload, type, oldProps, newProps) {
      // noop
    },
  },
});

export default PDFRenderer;
