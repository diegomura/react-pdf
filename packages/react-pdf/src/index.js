/* global Blob */
'use strict';

import { PDFRenderer, createElement } from './renderer';
import StyleSheet from './Stylesheet';

const View = 'VIEW';
const Text = 'TEXT';
const Page = 'PAGE';
const Document = 'DOCUMENT';

const pdf = input => {
  async function parse(input) {
    const result = input.render();

    if (input.props.onRender) {
      input.props.onRender();
    }

    return result;
  }

  async function toBlob() {
    const render = await parse(input);

    return new Blob([render], {
      type: 'application/pdf',
    });
  }

  async function toBuffer() {
    const render = await parse(input);
    return new Buffer(render);
  }

  async function toString() {
    const render = await parse(input);

    return render;
  }

  return {
    toBuffer,
    toBlob,
    toString,
  };
};

export {
  PDFRenderer,
  View,
  Text,
  Page,
  Document,
  StyleSheet,
  createElement,
  pdf,
};
