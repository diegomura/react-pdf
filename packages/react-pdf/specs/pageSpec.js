import React from 'react';
import { Document, Page } from '../src';
import MockDate from 'mockdate';
import render from './testRenderer';

describe('<Page />', () => {
  beforeEach(() => {
    MockDate.set(new Date('2016', '1', '1'));
  });

  const page = (props = {}) => (
    <Document>
      <Page {...props} />
    </Document>
  );

  test('Should render A4 blank portrait page by default', () => {
    render(page()).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  test('Should render given size', () => {
    render(page({ size: 'A0' })).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  test('Should render given background color', () => {
    render(
      page({
        style: {
          backgroundColor: 'tomato',
        },
      }),
    ).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  test('Should render landscape page', () => {
    render(page({ orientation: 'landscape' })).then(result => {
      expect(result).toMatchSnapshot();
    });
  });
});
