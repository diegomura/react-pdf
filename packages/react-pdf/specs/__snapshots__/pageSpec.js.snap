// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Page /> Should render given background color 1`] = `
"%PDF-1.7
%����
9 0 obj
<<
/Keywords ()
/CreationDate (D:20170422014110)
/Creator (react-pdf)
/Producer (react-pdf)
>>
endobj
10 0 obj
<<
/Type /Font
/Subtype /Type1
/Name /F1
/BaseFont /Helvetica
>>
endobj
11 0 obj
<<
/Type /Catalog
/Pages 12 0 R
>>
endobj
12 0 obj
<<
/Type /Pages
/Kids [13 0 R]
/Count 1
>>
endobj
13 0 obj
<<
/Type /Page
/Parent 12 0 R
/Contents [16 0 R 15 0 R ]
/Resources 14 0 R
/MediaBox [0 0 2383.93994140625 3370.389892578125]
>>
endobj
16 0 obj
<<
/Length 16
>>
stream
/DeviceRGB cs



endstream
endobj
15 0 obj
<<
/Length 31
>>
stream
1 0 0 -1 0 3370.389892578125 cm
endstream
endobj
14 0 obj
<<
/ProcSet [/PDF/Text]
/Font <</F1 10 0 R>>
>>
endobj
xref
0 9
0000000000 65535 f
0000000015 00000 n
0000000125 00000 n
0000000206 00000 n
0000000257 00000 n
0000000316 00000 n
0000000610 00000 n
0000000528 00000 n
0000000461 00000 n
trailer
<<
/Size 8
/Root 11 0 R
/Info 9 0 R
>>
startxref
674
%%EOF"
`;

exports[`<Page /> Should render given size 1`] = `
"%PDF-1.7
%����
1 0 obj
<<
/Keywords ()
/CreationDate (D:20170422014110)
/Creator (react-pdf)
/Producer (react-pdf)
>>
endobj
2 0 obj
<<
/Type /Font
/Subtype /Type1
/Name /F1
/BaseFont /Helvetica
>>
endobj
3 0 obj
<<
/Type /Catalog
/Pages 4 0 R
>>
endobj
4 0 obj
<<
/Type /Pages
/Kids [5 0 R]
/Count 1
>>
endobj
5 0 obj
<<
/Type /Page
/Parent 4 0 R
/Contents [8 0 R 7 0 R ]
/Resources 6 0 R
/MediaBox [0 0 595.280029296875 841.8900146484375]
>>
endobj
8 0 obj
<<
/Length 16
>>
stream
/DeviceRGB cs



endstream
endobj
7 0 obj
<<
/Length 31
>>
stream
1 0 0 -1 0 841.8900146484375 cm
endstream
endobj
6 0 obj
<<
/ProcSet [/PDF/Text]
/Font <</F1 2 0 R>>
>>
endobj
xref
0 9
0000000000 65535 f
0000000015 00000 n
0000000125 00000 n
0000000205 00000 n
0000000254 00000 n
0000000311 00000 n
0000000598 00000 n
0000000517 00000 n
0000000451 00000 n
trailer
<<
/Size 8
/Root 3 0 R
/Info 1 0 R
>>
startxref
660
%%EOF"
`;

exports[`<Page /> Should render landscape page 1`] = `
"%PDF-1.7
%����
17 0 obj
<<
/Keywords ()
/CreationDate (D:20170422014110)
/Creator (react-pdf)
/Producer (react-pdf)
>>
endobj
18 0 obj
<<
/Type /Font
/Subtype /Type1
/Name /F1
/BaseFont /Helvetica
>>
endobj
19 0 obj
<<
/Type /Catalog
/Pages 20 0 R
>>
endobj
20 0 obj
<<
/Type /Pages
/Kids [21 0 R]
/Count 1
>>
endobj
21 0 obj
<<
/Type /Page
/Parent 20 0 R
/Contents [24 0 R 23 0 R ]
/Resources 22 0 R
/MediaBox [0 0 595.280029296875 841.8900146484375]
>>
endobj
24 0 obj
<<
/Length 102
>>
stream
/DeviceRGB cs
1 0.38823529411764707 0.2784313725490196 scn
0 0 595.280029296875 841.8900146484375 re
f
endstream
endobj
23 0 obj
<<
/Length 31
>>
stream
1 0 0 -1 0 841.8900146484375 cm
endstream
endobj
22 0 obj
<<
/ProcSet [/PDF/Text]
/Font <</F1 18 0 R>>
>>
endobj
xref
0 9
0000000000 65535 f
0000000015 00000 n
0000000126 00000 n
0000000207 00000 n
0000000258 00000 n
0000000317 00000 n
0000000698 00000 n
0000000616 00000 n
0000000462 00000 n
trailer
<<
/Size 8
/Root 19 0 R
/Info 17 0 R
>>
startxref
762
%%EOF"
`;
